colcon_core/argument_default.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2020 Dirk Thomas
+# Licensed under the Apache License, Version 2.0
+
+
+class BooleanDefaultValue:
+    """Boolean value distinguishable from explicitly passed arguments."""
+
+    def __init__(self, value):  # noqa: D107
+        assert isinstance(value, bool)
+        self._value = value
+
+    def __bool__(self):  # noqa: D105
+        return self._value
+
+
+class ListDefaultValue(list):
+    """List value distinguishable from explicitly passed arguments."""
+
+    pass
+
+
+class StringDefaultValue(str):
+    """String value distinguishable from explicitly passed arguments."""
+
+    pass
+
+
+_types = {
+    bool: BooleanDefaultValue,
+    list: ListDefaultValue,
+    str: StringDefaultValue,
+}
+
+
+def wrap_default_value(value):
+    """
+    Wrap a default value in a custom type.
+
+    The custom type makes the default value distinguishable from explicitly
+    passed arguments independent of the value.
+    The custom types try to match the behavior of the original type as much as
+    possible.
+
+    :param value: The default value to be wrapped
+    :returns: The wrapped value if the value type is supported, otherwise the
+      passed value
+    """
+    global _types
+    if is_default_value(value):
+        assert False
```

Comment:
Why assert False?

Reply:
A value which is already wrapped shouldn't be wrapped again. I switched the assertion into an exception for better feedback: fafb10d229236bcdac988c06abee16cf8ac82791.